pkg/monitor/controller_test.go

```diff
@@ -0,0 +1,719 @@
+			expectedError:    "error creating agent crd as crd already exists",
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			clientset := fakeclientset.NewSimpleClientset()
+			if tt.existingAgentCRD != nil {
+				clientset = fakeclientset.NewSimpleClientset(tt.existingAgentCRD)
+			}
+			tt.prepareReactor(clientset)
+			controller := newControllerMonitor(clientset)
+			stopCh := make(chan struct{})
+			initController(controller, stopCh)
+			defer close(stopCh)
+			err := controller.controllerMonitor.createAgentCRD("testAgentCRD")
+			if tt.expectedError != "" {
+				assert.Equal(t, tt.expectedError, err.Error())
+			} else {
+				require.NoError(t, err)
+				crd, err := controller.controllerMonitor.client.CrdV1beta1().AntreaAgentInfos().Get(ctx, tt.expectedAgentCRD.Name, metav1.GetOptions{})
+				require.NoError(t, err)
+				assert.Equal(t, tt.expectedAgentCRD, crd)
+			}
+		})
+	}
+}
+
+func TestDeleteStaleAgentCRD(t *testing.T) {
+	tc := []struct {
+		name                 string
+		agentCRDs            []string
+		existingExternalNode *v1alpha1.ExternalNode
+		prepareReactor       func(clientset *fakeclientset.Clientset)
+		expectedError        string
+		remainingAgentCRDs   []string
+	}{
+		{
+			name:               "No external Nodes present",
+			agentCRDs:          []string{"testAgentCRD1", "testAgentCRD2", "testAgentCRD3", "testAgentCRD4"},
+			prepareReactor:     func(clientset *fakeclientset.Clientset) {},
+			remainingAgentCRDs: []string{},
+		},
+		{
+			name: "No Agent CRDs present",
+			prepareReactor: func(clientset *fakeclientset.Clientset) {
+				clientset.PrependReactor("list", "antreaagentinfos", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, &v1beta1.AntreaAgentInfoList{}, errors.New("error getting agent crds")
+				})
+			},
+			expectedError: "error getting agent crds",
+		},
+		{
+			name:      "Both Agent CRDs and external Node present",
+			agentCRDs: []string{"testAgentCRD1", "testAgentCRD2"},
+			existingExternalNode: &v1alpha1.ExternalNode{ObjectMeta: metav1.ObjectMeta{
+				Name:      "testAgentCRD1",
+				Namespace: "ns1",
+				Labels:    map[string]string{"en": "vm2"},
+			},
+			},
+			prepareReactor:     func(clientset *fakeclientset.Clientset) {},
+			remainingAgentCRDs: []string{"testAgentCRD1"},
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			clientset := fakeclientset.NewSimpleClientset()
+			if tt.existingExternalNode != nil {
+				clientset = fakeclientset.NewSimpleClientset(tt.existingExternalNode)
+			}
+			tt.prepareReactor(clientset)
+			controller := newControllerMonitor(clientset)
+			stopCh := make(chan struct{})
+			initController(controller, stopCh)
+			defer close(stopCh)
+			controller.controllerMonitor.externalNodeEnabled = true
+			for _, crd := range tt.agentCRDs {
+				agentCRD := new(v1beta1.AntreaAgentInfo)
+				agentCRD.Name = crd
+				agentCRD.ResourceVersion = "0"
```

Comment:
we usually use this syntax:
```go
agentCRD := &1beta1.AntreaAgentInfo{
        Name: crd,
        ResourceVersion: "0",
}
```

Reply:
This syntax gives error 'unknown field Name in struct literal' & 'unknown field ResourceVersion in struct literal'